scripts/flow/v11/flowPatcher.js

```diff
@@ -0,0 +1,101 @@
+import {FlowInfo, getTokenByIdOrActorId, processFlowInfo} from "../common.js";
+
+const _patchFlow = (flowId, patch) => {
+	const originalFlow = game.lancer.flowSteps.get(flowId);
+	if (!originalFlow) throw new Error(`Could not find flow "${flowId}" to patch!`);
+
+	game.lancer.flowSteps.set(flowId, async (...args) => {
+		await patch(...args);
+		return originalFlow(...args);
+	})
+};
```

Comment:
This does look like the right way to _extend_ (instead of replace) a flow step with new functionality. It now occurs to me that if multiple modules try to work with the same steps, there's no conflict resolution method. 🤔

Reply:
Yes, that is what made me think I was doing something wrong! It's a hard problem to solve, I wonder if we could use [libWrapper](https://github.com/ruipin/fvtt-lib-wrapper)? I will investigate tomorrow (GMT) as libWrapper has good conflict handling.

We don't really _need_ "extending" for Weapon FX though. A standard Foundry `Hooks.callAll( ... )` would work too. For example, if there were `lancer.preFlowStep<name>` and `lancer.postFlowStep<name>` hooks (and hopefully the `<name>` part could be dynamically generated by the flow-running logic) added to the system for external use?

Reply:
I could add hooks for the flow steps, that wouldn't be too difficult. 

The other option in LWFX's case would be to continue hooking on chat message creation, and instead of parsing the HTML, use the data in the `ChatMessage` document's flags. Most of the flow state gets sent there. (Eventual plan is to properly serialize all of the state so it can he used for e.g. reroll buttons.)

I do think using the flow step is a more elegant solution for what LWFX wants to do - it isn't actually related to chat stuff.

Reply:
Moving away from chat hooks is good, and it opens up more possibilities for future expansion. Flow `Hook`s would also be useful, not only for LWFX but for other modules. The hooks API is easy and familiar!

I tried out libWrapper, and it does not work as it cannot patch methods stored in a map. Would you consider switching the Lancer system from a map to assigning the methods as properties on an object?

Example:

```
// old
flowSteps.set("printGenericCard", printGenericCard);
// new
flowSteps.printGenericCard = printGenericCard;
```

This would enable flow patching of the following form:

```
// Copy existing function as an example
game.lancer.flowSteps.emptyStep = game.lancer.flowSteps.get("emptyStep")

// Add a wrapper - libWrapper will handle conflict resolution and user experience
libWrapper.register("lancer-weapon-fx", "game.lancer.flowSteps.emptyStep", (wrapped, ...args) => {
  console.warn("Custom wrapper ran!");
  return wrapped(...args);
}, "WRAPPER");
```

![image](https://github.com/Z3nner/lancer-weapon-fx/assets/9784519/f0fd8d0c-24fd-4ef4-a879-86b99735bb31)

I think enabling libWrapper would save a lot of pain in the long run, and enables familiar and well-proven tools for devs. Between hooks and libWrapper (almost) every use-case would be covered, which is boring, but hopefully easy!

Reply:
As an aside libWrapper was the most installed library in the [year-end review](https://foundryvtt.com/article/year-in-review-2024/) with nearly 75% of users having it installed!

Reply:
I have made a very rough attempt at the "hook" implementation here: https://github.com/Eranziel/foundryvtt-lancer/pull/652

Reply:
https://github.com/Eranziel/foundryvtt-lancer/blob/master/docs/flow_api.md#extending-flows

There is now a blessed way of inserting custom steps into flows.

Reply:
Bolts beat me to this comment. ;) The new hooks and flow registration API are available in Lancer 2.0.2, I'm very interested to hear how they are to work with from the module side.

Reply:
Very nice, thank you! I have updated the PR to use the new hooks.